pkg/blobaccess/file/access.go

```diff
@@ -0,0 +1,338 @@
+	_ bpi.BlobAccess   = (*fileBlobAccessView)(nil)
+	_ bpi.FileLocation = (*fileBlobAccessView)(nil)
+)
+
+func (f *fileBlobAccessView) Dup() (bpi.BlobAccess, error) {
+	b, err := f._blobAccess.Dup()
+	if err != nil {
+		return nil, err
+	}
+	return &fileBlobAccessView{b, f.access}, nil
+}
+
+func (f *fileBlobAccessView) FileSystem() vfs.FileSystem {
+	return f.access.fs
+}
+
+func (f *fileBlobAccessView) Path() string {
+	return f.access.path
+}
+
+func BlobAccessWithCloser(closer io.Closer, mime string, path string, fss ...vfs.FileSystem) bpi.BlobAccess {
+	fb := &fileBlobAccess{fileDataAccess{fs: utils.FileSystem(fss...), path: path}, mime}
+	return &fileBlobAccessView{
+		bpi.NewBlobAccessForBase(fb, closer),
+		&fb.fileDataAccess,
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type temporaryFileBlob struct {
+	_blobAccess
+	lock       sync.Mutex
+	path       string
+	file       vfs.File
+	filesystem vfs.FileSystem
+}
+
+var (
+	_ bpi.BlobAccessBase = (*temporaryFileBlob)(nil)
+	_ bpi.FileLocation   = (*temporaryFileBlob)(nil)
+)
+
+func (b *temporaryFileBlob) Validate() error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	if b.path == "" {
+		return bpi.ErrClosed
+	}
+	ok, err := vfs.Exists(b.filesystem, b.path)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return errors.ErrNotFound("file", b.path)
+	}
+	return nil
+}
+
+func (b *temporaryFileBlob) Close() error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	if b.path != "" {
+		list := errors.ErrListf("temporary blob")
+		if b.file != nil {
+			list.Add(b.file.Close())
+		}
+		list.Add(b.filesystem.Remove(b.path))
+		b.path = ""
+		b.file = nil
+		b._blobAccess = nil
+		return list.Result()
+	}
+	return nil
+}
+
+func (b *temporaryFileBlob) FileSystem() vfs.FileSystem {
+	return b.filesystem
+}
+
+func (b *temporaryFileBlob) Path() string {
+	return b.path
+}
+
+func ForTemporaryFile(mime string, temp vfs.File, fss ...vfs.FileSystem) bpi.BlobAccess {
+	return bpi.NewBlobAccessForBase(&temporaryFileBlob{
+		_blobAccess: BlobAccess(mime, temp.Name(), fss...),
+		filesystem:  utils.FileSystem(fss...),
+		path:        temp.Name(),
+		file:        temp,
+	})
+}
+
+func ForTemporaryFileWithMeta(mime string, digest digest.Digest, size int64, temp vfs.File, fss ...vfs.FileSystem) bpi.BlobAccess {
+	return bpi.NewBlobAccessForBase(bpi.BaseAccessForDataAccessAndMeta(mime, &temporaryFileBlob{
+		_blobAccess: BlobAccess(mime, temp.Name(), fss...),
+		filesystem:  utils.FileSystem(fss...),
+		path:        temp.Name(),
+		file:        temp,
+	}, digest, size))
+}
+
+func ForTemporaryFilePath(mime string, temp string, fss ...vfs.FileSystem) bpi.BlobAccess {
+	return bpi.NewBlobAccessForBase(&temporaryFileBlob{
+		_blobAccess: BlobAccess(mime, temp, fss...),
+		filesystem:  utils.FileSystem(fss...),
+		path:        temp,
+	})
+}
+
+func ForTemporaryFilePathWithMeta(mime string, digest digest.Digest, size int64, temp string, fss ...vfs.FileSystem) bpi.BlobAccess {
```

Comment:
What is `Meta`?
Can't we stick to the super fancy `OptionUtils` pattern?

Reply:
The signature for these functions cannot be changed without being incompatible.

Reply:
same here